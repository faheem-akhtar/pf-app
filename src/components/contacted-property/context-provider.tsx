import { FunctionComponent } from 'react';

import { apiContactedPropertiesCreateFetcher } from 'api/contacted-properties/create/fetcher';
import { useApiContactedProperties } from 'api/contacted-properties/hook';
import { useUserPropertyState } from 'components/user-property/state.hook';
import { dateToIso } from 'helpers/date/to-iso';

import { ContactedPropertyContext } from './context';
import { ContactedPropertyContextInterface } from './context.interface';
import { ContactedPropertyInterface } from './interface';
import { CONTACTED_PROPERTY_STORAGE_KEY } from './storage-key.constant';

export const ContactedPropertyContextProvider: FunctionComponent = ({ children }) => {
  const contactedPropertiesResponse = useApiContactedProperties();
  const [contactedProperties, setContactedProperty] = useUserPropertyState<ContactedPropertyInterface>(
    CONTACTED_PROPERTY_STORAGE_KEY,
    contactedPropertiesResponse,
    ({ propertyId, contactType, contactDate }) =>
      ({ propertyId, contactType, contactDate } as ContactedPropertyInterface),
    apiContactedPropertiesCreateFetcher
  );

  const value: ContactedPropertyContextInterface = {
    data: contactedProperties,
    add: (propertyId, contactType) => {
      const property = {
        propertyId,
        contactType,
        contactDate: dateToIso(new Date()),
      } as ContactedPropertyInterface;

      setContactedProperty(property);
    },
  };

  return <ContactedPropertyContext.Provider value={value}>{children}</ContactedPropertyContext.Provider>;
};
