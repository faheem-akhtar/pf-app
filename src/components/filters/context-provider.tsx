import { NextRouter, Router, useRouter } from 'next/dist/client/router';

import { filtersValueToQuery } from './value/to-query';
import { urlQuerySerialize } from 'helpers/url-query/serialize';
import { useFiltersValueState } from './value/state.hook';

import { FiltersContext } from './context';
import { FiltersContextInterface } from './context.interface';
import { FiltersContextProviderPropsInterface } from './context-provider-props.interface';
import { FiltersParametersEnum } from 'enums/filters/parameters.enum';
import { FiltersValueInterface } from './value/interface';
import { UrlQueryType } from 'types/url/query.type';
import { useEffect } from 'react';

/**
 * Change browser url path
 */
const navigate = (router: NextRouter, filtersValue: FiltersValueInterface): void => {
  const query = filtersValueToQuery(filtersValue);
  const url = `${router.pathname}?${urlQuerySerialize(query as UrlQueryType)}`;

  if (router.asPath !== url) {
    router.push(url);
  }
};

export const FiltersContextProvider = ({
  filtersValueFromQuery,
  filtersData,
  children,
}: FiltersContextProviderPropsInterface): JSX.Element => {
  const router = useRouter();

  const { filtersValue, changeFiltersValue, setFiltersValue, resetFiltersValue, filtersValueIsDefault } =
    useFiltersValueState(filtersData, filtersValueFromQuery);

  const updateFiltersValue: FiltersContextInterface['change'] = (updater) => {
    const updatedFiltersValue = updater(filtersValue);
    const nextFiltersValue = changeFiltersValue(updatedFiltersValue);
    navigate(router, nextFiltersValue);
    return nextFiltersValue;
  };

  const value = {
    value: filtersValue,
    valueIsDefault: filtersValueIsDefault,
    data: filtersData,
    change: updateFiltersValue,
    set: (nextFiltersValue: FiltersValueInterface): void => {
      setFiltersValue(nextFiltersValue);
      navigate(router, nextFiltersValue);
    },
    reset: (): void => {
      const nextFiltersValue = resetFiltersValue();
      navigate(router, nextFiltersValue);
    },
  };

  useEffect(() => {
    const onRouteChangeComplete = (url: string): void => {
      const queryPage = parseInt(url.match(/page=(\d+)/i)?.[1] || '1', 10);

      if (filtersValue[FiltersParametersEnum.pageNumber] !== queryPage) {
        setFiltersValue({ ...filtersValue, [FiltersParametersEnum.pageNumber]: queryPage });
      }
    };

    Router.events.on('routeChangeComplete', onRouteChangeComplete);

    return (): void => Router.events.off('routeChangeComplete', onRouteChangeComplete);
  });

  return <FiltersContext.Provider value={value}>{children}</FiltersContext.Provider>;
};
