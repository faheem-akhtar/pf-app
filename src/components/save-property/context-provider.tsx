import { FunctionComponent } from 'react';

import { useSavedPropertiesState } from 'components/saved-properties/state.hook';

import { SavePropertyContext } from './context';
import { SavePropertyContextInterface } from './context.interface';

export const SavePropertyContextProvider: FunctionComponent = ({ children }) => {
  const [savedProperties, setSavedProperty, removeSavedProperty] = useSavedPropertiesState();

  const value: SavePropertyContextInterface = {
    propertyIds: savedProperties,
    toggle: (propertyId) => {
      const propId = parseInt(propertyId, 10);
      const isSaved = savedProperties.includes(propId);
      if (isSaved) {
        removeSavedProperty(propId);
      } else {
        setSavedProperty(propId);
      }
    },
  };

  return <SavePropertyContext.Provider value={value}>{children}</SavePropertyContext.Provider>;
};
