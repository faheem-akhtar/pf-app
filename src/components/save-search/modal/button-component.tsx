import { useContext, useEffect, useRef, useState } from 'react';

import { AuthModalComponent } from 'components/auth/modal/component';
import { IconThinStarTemplate } from 'components/icon/thin/star-template';
import { ModalComponent } from 'components/modal/component';
import { SnackbarContext } from 'components/snackbar/context';
import { UserContext } from 'context/user/context';
import { useTranslation } from 'helpers/translation/hook';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { OnBoardingComponent } from 'library/on-boarding/component';
import { OnBoardingTypeEnum } from 'library/on-boarding/type.enum';
import { AuthService } from 'services/auth/service';
import { AuthSubscribeEventTypeEnum } from 'services/auth/subscribe-event-type.enum';
import { AuthSubscriberType } from 'services/auth/subscriber.type';

import { SaveSearchContext } from '../context';
import { saveSearchTracker } from '../tracker';
import { SaveSearchModalContentComponent } from './content-component';
import styles from './save-search-modal-component.module.scss';

export const SaveSearchModalButtonComponent = ({ visibleTooltip }: { visibleTooltip: boolean }): JSX.Element => {
  const { t } = useTranslation();
  const user = useContext(UserContext);
  const snackbar = useContext(SnackbarContext);
  const saveSearch = useContext(SaveSearchContext);
  const [autoOpenDialog, setAutoOpenDialog] = useState(false);
  const openFiltersRef = useRef<() => void>(() => null);
  const closeFiltersRef = useRef<() => void>(() => null);
  const openLoginRef = useRef<() => void>(() => null);
  const closeLoginRef = useRef<() => void>(() => null);

  const openDialog = (): void => {
    if (user) {
      openFiltersRef.current();
      saveSearchTracker.onOpenCreateDialog();
    } else {
      openLoginRef.current();
      setAutoOpenDialog(true);
    }
  };

  useEffect(() => {
    if (autoOpenDialog && user) {
      closeLoginRef.current();
      if (saveSearch.filtered.length === 0) {
        openDialog();
      } else {
        snackbar.alert({ message: t('save_search/exist-notification') });
      }
      setAutoOpenDialog(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [autoOpenDialog, user, saveSearch.filtered]);

  useEffect(() => {
    if (autoOpenDialog) {
      const onUpdate: AuthSubscriberType = (user, meta) => {
        if (user && meta?.eventType === AuthSubscribeEventTypeEnum.login) {
          saveSearchTracker.onSignInSuccess();
        }
        if (user && meta?.eventType === AuthSubscribeEventTypeEnum.register) {
          saveSearchTracker.onSignUpSuccess();
        }
      };
      const unsubscribeAuth = AuthService.subscribe(onUpdate);

      return (): void => unsubscribeAuth();
    }
  }, [autoOpenDialog]);

  return (
    <>
      <div className={styles.tooltipWrapper}>
        <ButtonTemplate
          type='button'
          componentType={ButtonComponentTypeEnum.tertiary}
          size={ButtonSizeEnum.small}
          onClick={(): void => {
            if (saveSearch.filtered.length === 0) {
              openDialog();
            }
            saveSearchTracker.onClickCta();
          }}
          icon={{
            component: IconThinStarTemplate,
            position: ButtonIconPositionEnum.left,
            className: saveSearch.filtered.length ? styles.icon__active : styles.icon__inactive,
          }}
        >
          {t('save-search/cta-label')}
        </ButtonTemplate>
        {visibleTooltip && (
          <OnBoardingComponent
            name={OnBoardingTypeEnum.saveSearchTooltip}
            className={styles.tooltip}
            arrowClassName={styles.tooltipArrow}
            onClose={(): void => {
              saveSearchTracker.onTooltipOpen();
              saveSearchTracker.onTooltipClose();
            }}
          >
            <>
              <p className={styles.tooltipTitle}>{t('on-boarding/save-search-title')}</p>
              {t('on-boarding/save-search-description')}
            </>
          </OnBoardingComponent>
        )}
      </div>
      <ModalComponent openRef={openFiltersRef} closeRef={closeFiltersRef}>
        <SaveSearchModalContentComponent close={(): void => closeFiltersRef.current()} />
      </ModalComponent>
      <ModalComponent openRef={openLoginRef} closeRef={closeLoginRef} overlay>
        <AuthModalComponent
          cancel={(): void => {
            closeLoginRef.current();
            setAutoOpenDialog(false);
          }}
        />
      </ModalComponent>
    </>
  );
};
