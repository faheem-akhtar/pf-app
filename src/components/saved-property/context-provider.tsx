import { FunctionComponent } from 'react';

import { apiSavedPropertiesCreateFetcher } from 'api/saved-properties/create-fetcher';
import { useApiSavedProperties } from 'api/saved-properties/hook';
import { apiSavedPropertiesRemoveFetcher } from 'api/saved-properties/remove-fetcher';
import { useUserPropertyState } from 'components/user/property/state.hook';
import { dateToIso } from 'helpers/date/to-iso';

import { SavedPropertyContext } from './context';
import { SavedPropertyContextInterface } from './context.interface';
import { SavedPropertyInterface } from './interface';
import { SAVED_PROPERTY_STORAGE_KEY } from './storage-key.constant';

export const SavedPropertyContextProvider: FunctionComponent = ({ children }) => {
  const savedPropertiesResponse = useApiSavedProperties();
  const [savedProperties, setSavedProperty, removeSavedProperty] = useUserPropertyState<SavedPropertyInterface>(
    SAVED_PROPERTY_STORAGE_KEY,
    savedPropertiesResponse,
    ({ propertyId, saveDate }) => ({ propertyId, saveDate } as SavedPropertyInterface),
    apiSavedPropertiesCreateFetcher,
    apiSavedPropertiesRemoveFetcher
  );

  const value: SavedPropertyContextInterface = {
    data: savedProperties,
    toggle: (propertyId) => {
      const property = {
        propertyId: parseInt(propertyId, 10),
        saveDate: dateToIso(new Date()),
      } as SavedPropertyInterface;
      const isSaved = !!savedProperties.find((item) => item.propertyId === property.propertyId);

      if (isSaved) {
        removeSavedProperty(property);
      } else {
        setSavedProperty(property);
      }
    },
  };

  return <SavedPropertyContext.Provider value={value}>{children}</SavedPropertyContext.Provider>;
};
