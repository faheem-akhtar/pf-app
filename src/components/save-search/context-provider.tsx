import { FunctionComponent, useContext, useEffect, useState } from 'react';

import { apiSaveSearchCreateFetcher } from 'api/save-search/create/fetcher';
import { useApiSaveSearch } from 'api/save-search/hook';
import { FiltersContext } from 'components/filters/context';
import { UserContext } from 'components/user/context';
import { FiltersParametersEnum } from 'enums/filters/parameters.enum';

import { SaveSearchContext } from './context';
import { SaveSearchContextInterface } from './context.interface';
import { saveSearchFilterEquality } from './filter-equality';
import { saveSearchFiltersMapper } from './filters-mapper';
import { SaveSearchInterface } from './interface';
import { saveSearchTracker } from './tracker';

export const SaveSearchContextProvider: FunctionComponent = ({ children }) => {
  const user = useContext(UserContext);
  const saveSearchResponse = useApiSaveSearch();
  const saveSearchResponseData = saveSearchResponse.ok ? saveSearchResponse.data : [];
  const [searches, setSearches] = useState<SaveSearchInterface[]>([]);
  const filtersCtx = useContext(FiltersContext);
  const filters = saveSearchFiltersMapper(filtersCtx?.value || {});

  useEffect(() => {
    if (user && saveSearchResponse.ok) {
      setSearches(saveSearchResponseData);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user, saveSearchResponse.ok]);

  const value: SaveSearchContextInterface = {
    ok: saveSearchResponse.ok,
    data: searches,
    filtered: searches.filter((item) => saveSearchFilterEquality(item.filters, filters)),
    create: ({ name, frequency }) =>
      apiSaveSearchCreateFetcher({ name, frequency, filters: filtersCtx.value }).then((response) => {
        if (response.ok) {
          setSearches([...searches, response.data]);
          saveSearchTracker.onCreateSuccess(
            filtersCtx.value[FiltersParametersEnum.locationsIds].map((location) => location.id)
          );
        }
        return response;
      }),
  };

  return <SaveSearchContext.Provider value={value}>{children}</SaveSearchContext.Provider>;
};
