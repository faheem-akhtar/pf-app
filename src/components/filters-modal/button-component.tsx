import { useRef } from 'react';

import { useTranslationHook } from 'helpers/hook/translation.hook';

import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { FiltersModalContentComponent } from './content-component';
import { IconThinFilterOpenTemplate } from 'components/icon/thin/filter-open-template';
import { ModalComponent } from 'components/modal/component';

// TODO-FE[CX-423] add tests
export const FiltersModalButtonComponent = (): JSX.Element => {
  const openFiltersRef = useRef<() => void>(() => null);
  const closeFiltersRef = useRef<() => void>(() => null);
  const { t } = useTranslationHook();

  // TODO-FE[TPNX-3152] display number of non default filters
  const nonDefaultFiltersCount = 0;

  return (
    <>
      <ButtonTemplate
        type='button'
        componentType={ButtonComponentTypeEnum.tertiary}
        size={ButtonSizeEnum.small}
        onClick={(): void => openFiltersRef.current()}
        icon={{ component: IconThinFilterOpenTemplate, position: ButtonIconPositionEnum.left }}
      >
        {t('filters')} <span>{nonDefaultFiltersCount ? `(${nonDefaultFiltersCount})` : ''}</span>
      </ButtonTemplate>
      <ModalComponent openRef={openFiltersRef} closeRef={closeFiltersRef}>
        <FiltersModalContentComponent close={(): void => closeFiltersRef.current()} />
      </ModalComponent>
    </>
  );
};
