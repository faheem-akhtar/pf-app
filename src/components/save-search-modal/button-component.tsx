import { useRef, useState } from 'react';

import { useTranslationHook } from 'helpers/hook/translation.hook';

import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { IconThinStarTemplate } from 'components/icon/thin/star-template';
import { ModalComponent } from 'components/modal/component';

import styles from './save-search-modal-component.module.scss';

export const SaveSearchModalButtonComponent = (): JSX.Element => {
  const openFiltersRef = useRef<() => void>(() => null);
  const closeFiltersRef = useRef<() => void>(() => null);
  const { t } = useTranslationHook();

  const [isActive, setIsActive] = useState(false);

  return (
    <>
      <ButtonTemplate
        type='button'
        componentType={ButtonComponentTypeEnum.tertiary}
        size={ButtonSizeEnum.small}
        onClick={(): void => setIsActive(!isActive)}
        icon={{
          component: IconThinStarTemplate,
          position: ButtonIconPositionEnum.left,
          className: isActive ? styles.icon__active : styles.icon__inactive,
        }}
      >
        {t('save-search-cta')}
      </ButtonTemplate>
      <ModalComponent openRef={openFiltersRef} closeRef={closeFiltersRef}>
        TODO-FE[TPNX-3151] Save search
      </ModalComponent>
    </>
  );
};
