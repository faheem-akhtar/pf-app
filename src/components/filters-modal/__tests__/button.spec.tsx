import { act, render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { mockModalEnv } from 'mocks/modal-env/mock';
import { mockReactUseSwr } from 'mocks/react/use-swr.mock';
import { filtersContextPropsStub } from 'stubs/filters/context-props.stub';

import { FiltersContextProvider } from 'components/filters/context-provider';

import { FiltersModalButtonComponent } from '../button-component';

describe('FiltersModalButtonComponent', () => {
  const openFiltersRef = { current: jest.fn() };
  const closeFiltersRef = { current: jest.fn() };

  let renderResult: RenderResult;
  beforeEach(() => {
    mockReactUseSwr('api_user', {
      ok: true,
      data: {
        'save-search-tooltip': true,
      },
    });
    mockModalEnv();
    renderResult = render(
      <FiltersContextProvider {...filtersContextPropsStub()}>
        <FiltersModalButtonComponent visibleTooltip={false} />
      </FiltersContextProvider>
    );
  });

  it('should render without throwing any errors', () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it('should call FiltersModalContentComponent modal', () => {
    userEvent.click(screen.getByRole('button', { name: 'filters' }));
    act(openFiltersRef.current);

    expect(openFiltersRef.current).toHaveBeenCalledTimes(1);

    const titleElement = screen.getByRole('heading', { name: 'filters' });

    expect(titleElement).toBeInTheDocument();

    userEvent.click(titleElement);
    act(closeFiltersRef.current);

    expect(titleElement).not.toBeInTheDocument();
  });

  it('should not appear on-boarding tooltip', () => {
    expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
  });
});
