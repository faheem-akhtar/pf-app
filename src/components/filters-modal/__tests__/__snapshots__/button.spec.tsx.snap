// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersModalButtonComponent should render without throwing any errors 1`] = `
<div>
  <div
    class="tooltipWrapper"
  >
    <button
      class="button tertiary small filter_button"
    >
      <svg
        class="icon icon--left"
        height="24"
        viewBox="-5 -7 24 24"
        width="24"
      >
        <g
          id="svg-filter"
          stroke="none"
        >
          <path
            d="M6.565 4c-.083.326-.085.67 0 1H.5a.5.5 0 0 1 0-1h6.065zm2.01 4l.656 1H.5a.5.5 0 0 1 0-1h8.075zM.5 0h11.997a.5.5 0 0 1 0 1H.5a.5.5 0 0 1 0-1zm11.1 9.8l-3.498-5a.5.5 0 0 1 .4-.8H15.5a.5.5 0 0 1 .4.8l-3.5 5a.5.5 0 0 1-.8 0z"
          />
        </g>
      </svg>
      filters
       
      <span />
    </button>
  </div>
</div>
`;
